tests/test_operator.py

```diff
@@ -644,16 +644,18 @@ def test_tensor(self, func1):
     def test_complex(self, dtype):
```

Comment:
The introduction of an override is a good idea here.

I think this wants separating out into a couple of tests:
* Test that the correct expressions are generated
* Test that you can specify derivatives of complex fields
* Test that you can override complex values
* Test that you can mix complex and real symbols on the RHS

Note: do we also need to test that halo exchanges work with complex?

Reply:
Ok, will add some tests for those points.

Re: halo exchanges, I'll get back to you when I figure out what a halo exchange is

Reply:
A halo exchange is where each MPI rank swaps the data along its edges with its neighbours before advancing to the next timestep. It would probably worth checking that the MPI communication doesn't garble the data somehow when moving it between ranks. You would probably need some equations like:
```
from devito import Grid, TimeFunction, Eq, Operator

grid = Grid(shape=(4, 4))
x, y = grid.dimensions
t = grid.stepping_dim

u = TimeFunction(name='u', grid=grid)
v = TimeFunction(name='v', grid=grid)

eqns = [Eq(u[t+1, x, y], 1+sympy.I), Eq(v[t+1, x, y], sympy.I*u[t, x, y+1])]

op = Operator(eqns)
```

Reply:
Ok. I think what I'll do is leave these general complex arithmetic tests where they are (in `test_operator` and `test_gpu_common`) and add these lower-level ones to the `test_dtypes` module. Except this halo exchange one goes in `test_mpi` I guess